src/lib/pubkey/hss_lms/hss.cpp

```diff
@@ -0,0 +1,418 @@
+      size_t layer_h = layer_params.lms_params().h();
+      q.at(layer.get()) =
+         checked_cast_to<LMS_Tree_Node_Idx>(hss_idx.get() % checked_cast_to<uint64_t>(1ULL << layer_h));
+      hss_idx = hss_idx >> layer_h;
+   }
+   BOTAN_ARG_CHECK(hss_idx == HSS_Sig_Idx(0), "HSS Tree is exhausted");
+
+   return q;
+}
+
+}  // namespace
+
+HSS_LMS_Params::HSS_LMS_Params(std::vector<LMS_LMOTS_Params_Pair> lm_lmots_params) :
+      m_lms_lmots_params(std::move(lm_lmots_params)), m_max_sig_count(calc_max_sig_count()) {
+   BOTAN_ARG_CHECK(!m_lms_lmots_params.empty() && m_lms_lmots_params.size() <= HSS_MAX_LEVELS,
+                   "Invalid number of levels");
+}
+
+HSS_LMS_Params::HSS_LMS_Params(std::string_view algo_params) {
+   SCAN_Name scan(fmt("HSS-LMS({})", algo_params));
+
+   BOTAN_ARG_CHECK(scan.arg_count() >= 2 && scan.arg_count() <= HSS_MAX_LEVELS + 1, "Invalid number of arguments");
+   std::string hash = scan.arg(0);
+   BOTAN_ARG_CHECK(is_supported_hash_function(hash), "Supported HSS-LMS hash function");
+
+   for(size_t i = 1; i < scan.arg_count(); ++i) {
+      SCAN_Name scan_layer(scan.arg(i));
+      BOTAN_ARG_CHECK(scan_layer.algo_name() == "HW", "Invalid name for layer parameters");
+      BOTAN_ARG_CHECK(scan_layer.arg_count() == 2, "Invalid number of layer parameters");
+      const auto h =
+         checked_cast_to_or_throw<uint8_t, Invalid_Argument>(scan_layer.arg_as_integer(0), "Invalid parameter");
+      const auto w =
+         checked_cast_to_or_throw<uint8_t, Invalid_Argument>(scan_layer.arg_as_integer(1), "Invalid parameter");
+      m_lms_lmots_params.push_back({LMS_Params::create_or_throw(hash, h), LMOTS_Params::create_or_throw(hash, w)});
+   }
+   m_max_sig_count = calc_max_sig_count();
+}
+
+HSS_Sig_Idx HSS_LMS_Params::calc_max_sig_count() const {
+   uint32_t total_height_counter = 0;
+   for(HSS_Level level(0); level < L(); level++) {
+      total_height_counter += params_at_level(level).lms_params().h();
+   }
+   if(total_height_counter >= sizeof(HSS_Sig_Idx) * 8) {
+      return HSS_Sig_Idx(std::numeric_limits<HSS_Sig_Idx::wrapped_type>::max());
+   }
+   return HSS_Sig_Idx(1) << total_height_counter;
+}
+
+HSS_LMS_PrivateKeyInternal::HSS_LMS_PrivateKeyInternal(const HSS_LMS_Params& hss_params, RandomNumberGenerator& rng) :
+      m_hss_params(hss_params), m_current_idx(0), m_sig_size(HSS_Signature::size(m_hss_params)) {
+   m_hss_seed = rng.random_vec<LMS_Seed>(m_hss_params.params_at_level(HSS_Level(0)).lms_params().m());
+   m_identifier = rng.random_vec<LMS_Identifier>(LMS_IDENTIFIER_LEN);
+}
+
+std::shared_ptr<HSS_LMS_PrivateKeyInternal> HSS_LMS_PrivateKeyInternal::from_bytes_or_throw(
+   std::span<const uint8_t> key_bytes) {
+   if(key_bytes.size() < sizeof(HSS_Level) + sizeof(HSS_Sig_Idx)) {
+      throw Decoding_Error("Too few private key bytes.");
+   }
+   BufferSlicer slicer(key_bytes);
+
+   const auto L = load_be<HSS_Level>(slicer.take<sizeof(HSS_Level)>());
```

Comment:
This, being so verbose, is bothering me quite a bit, tbh.

For future work (and related to https://github.com/randombit/botan/pull/3716#discussion_r1576194649): We might want to consider allowing to pass `BufferSlicer` into a `load_be/le()` in-lieu of a "contiguous-range". Conceptually, they're almost the same anyway. It would be quite handy, particularly when working with strongly-typed integers. Many locations in this file could be simplified along those lines:

```suggestion
   const auto L = load_be<HSS_Level>(slicer);
```

... I'm on the fence to propose the same for `store_be/le()`. There, we can do `stuffer.append(store_be(someval))` which I find acceptable. For the record, the new overload would allow `store_be(stuffer, someval)` which might even be less readable. Regardless, a stuffer-based overload would also allow things like multi-input stores, e.g. `store_le(stuffer, val1, val2, val3, val4)` or `store_be(stuffer, array_of_vals)`.

(I'll create a ticket for it, if there's no objection. No reason to hold back on merging because of that, IMO.)

Reply:
For the record: https://github.com/randombit/botan/issues/4068